Sever dependency on Cargo in release mode

XrXr opened the pull request:
We want to build offline in release mode and currently `cargo --offline`
does not do what we want. In release mode, run `rustc` directly and
don't check for Cargo in `configure`. Reword the build prompt to include
the current build mode.

No CI changes yet, but this can be used to check that we build without
any external crates. For release mode users it's potentially a boon to
not need to install Cargo.

XrXr committed:
Sever dependency on Cargo in release mode

We want to build offline in release mode and currently `cargo --offline`
does not do what we want. In release mode, run `rustc` directly and
don't check for Cargo in `configure`. Reword the build prompt to include
the current build mode.

No CI changes yet, but this can be used to check that we build without
any external crates. For release mode users it's potentially a boon to
not need to install Cargo.

maximecb commented:
I support this but I do think we need to add something to a CI workflow to at least just check that we can configure and build the full `ruby` in release mode. We want to know if this becomes broken.

XrXr committed:
Test release mode on CI

XrXr commented:
We run 6 checks upstream in YJIT Ubuntu and we might want to reduce the number of checks we run before upstreaming. Matching the upstream number seems most reasonable.

We swap the C assembler test with `cargo test`. No change there.

XrXr committed:
Like in C YJIT, fill stats hash with code size info

In C YJIT, YJIT.runtime_stats always returns a hash even in non stats
builds. Release mode CI caught this difference in test-all.

XrXr committed:
Always verify blockid to fix unused warning

I was going to use debug_assert but the docs convinced me otherwise.
It always asserts in C YJIT so this matches.

maximecb commented on .github/workflows/yjit-ubuntu.yml:
```diff
@@ -47,6 +47,7 @@ jobs:
         ]
         configure: [
           "--enable-yjit=dev",
+          "--enable-yjit",
```
I agree that we're running a lot of tests and we don't necessary need to rerun all the tests in release mode.

I was suggesting we could just test that we are able to run `./configure --enable-yjit` and then build the full ruby, and then just run btest or something. Might be easier to do in a new CI workflow?

XrXr replied:
Well, I do think it's important to run all the test in release mode because it's the configuration users use. We run a mix of `-DRUBY_DEBUG` and non-debug builds on upstream CI and maybe we should inherit that mix? Subbing non-release builds with Rust YJIT release builds.

<img width="708" alt="Screen Shot 2022-04-05 at 15 27 25" src="https://user-images.githubusercontent.com/6457510/161834214-dba2d6b6-f1e1-4b93-9db2-66112792185d.png">

Also maybe swap out release threshold=1 with the bundler tests.

maximecb replied:
threshold=1 seems important because it helps catch more bugs, so we should definitely run at least `make check` with threshold=1

There's some redundancy in the current setup because `make check` includes `test-all` AFAIK.

Maybe we could run just `make check` in release mode. There's relatively pretty few bugs that would show up in release but not in debug.

XrXr replied:
Yeah, `make check` is btest+test-all+test-spec. So it seems upstream's setup is reasonable. Maybe at some point we want to setup a build farm outside the ruby org so we are not limited to just 6 configs.

maximecb commented:
Seeing some odd CI failures :O

XrXr committed:
Allow YJIT.insns_compiled in release mode for tests

It doesn't need Capstone and we can give it out easily. We already have
some tests that depend on this.
